typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,1016 @@
+
+In 1972, Per Martin-Löf found a type theory corresponding to predicate logic by introducing dependent types.
+This theory became known as intuitionistic type theory or commonly Martin-Löf type theory.
+
+In 1986, Thierry Coquand and Gérard Huet created the calculus of constructions, a dependent type theory for functions.
+This theory became the basis for Coq and Lean proof assistants.
+
+Since the early 2000s, type theory research has mostly focused on the homotopy type theory (HoTT).
+In HoTT the notion of equivalence is generalized to homotopy -- i.e. the topological notion of similarity.
+HoTT turns out to be powerful enough to discuss the foundations of mathematics again and is used in the univalent foundations project, which is arguably somewhat of a spiritual successor to Hilbert's program.
+
+## Simply typed lambda calculus
+
+Untyped lambda calculus contains some "puns". In the previous article, we noted that the term $fls$ (denoting boolean zero in Church encoding) and the term $c_0$ (denoting natural zero in Church encoding) have the same representation.
+
+Furthermore, untyped lambda calculus allows for constructions like, for example, $plus\; tru\; c_2,$ which don't have a defined semantic meaning.
+
+Various typed lambda calculi introduce some restrictions on the form that terms in different expressions might have, as to avoid this punning and meaningless terms.
+
+### Lambda calculus with atomic boolean values
+
+To facilitate the discussion, let us introduce a couple additional atomic value-terms to our lambda calculus syntax, namely, values $true$ and $false$.
+This, of course, makes our lambda calculus impure.
+
+The reason we need to do this will become apparent later, but for now, suffice it to say that we can't have simply typed pure lambda calculus as the "simple" type system is not expressive enough.
+
+Since we introduced atomic values, we also need to introduce an eliminator (i.e. consumer) for these values, in this case let's consider an $\mathbf{if} \ldots \mathbf{then} \ldots \mathbf{else} \ldots$ operator.
+
+The semantics of this operator should be reasonably obvious.
+We'll only note that there are two meaningful options here: either both branches are computed (corresponding to "eager" evaluation) or only the matching one is (corresponding to "lazy" evaluation).
+In practice, programming languages generally prefer the second option for obvious reasons, but the exact choice doesn't affect the following discussion.
+
+So, we'd like to avoid meaningless terms.
+What terms would be meaningful in this context?
+For one, we expect the first argument to be either $true$ or $false$.
+For two, we'd like both branches to have the same general "form" to avoid complications with nesting this construct into other computations.
+
+Formally speaking, "meaningless" terms are called "stuck" -- the term is not a value with defined semantic meaning, and further reduction is impossible.
```

Comment:
Now you say "stuck", but you do not explain what "stuck" means. Really, this is something you should've started with to motivate introducing a type system in the first place. We want to avoid stuck terms.